WalletWasabi/WabiSabi/Client/CoinJoinClient.cs

```diff
@@ -43,9 +43,35 @@ public class CoinJoinClient
 		public KeyManager Keymanager { get; }
 		private RoundStateUpdater RoundStatusUpdater { get; }
 
-		public async Task StartCoinJoinAsync(CancellationToken cancellationToken, IEnumerable<Money>? forcedOutputDenominations = null)
+		public async Task<bool> StartCoinJoinAsync(CancellationToken cancellationToken, IEnumerable<Money>? forcedOutputDenominations = null)
+		{
+			var currentRoundState = await RoundStatusUpdater.CreateRoundAwaiter(roundState => roundState.Phase == Phase.InputRegistration, cancellationToken).ConfigureAwait(false);
+
+			// This should be roughly log(#inputs), it could be set slightly
+			// higher if more inputs are observed but that involves trusting the
+			// coordinator with those values. Therefore, conservatively set this
+			// so that a maximum of 5 blame rounds are executed.
+			// FIXME should smaller rounds abort earlier?
+			var tryLimit = 6;
+
+			for (var tries = 0; tries < tryLimit; tries++)
+			{
+				if (await StartRoundAsync(currentRoundState, cancellationToken, forcedOutputDenominations))
+				{
+					return true;
+				}
+				else
+				{
+					var blameRoundState = await RoundStatusUpdater.CreateRoundAwaiter(roundState => roundState.BlameOf == currentRoundState.Id, cancellationToken).ConfigureAwait(false);
```

Comment:
I think that perhaps this should have a different cancellation token source, a linked one; otherwise it could wait forever.

Reply:
:+1:

is this a general thing, should cancellation tokens be linked instead of shared?

Reply:
You need a linked cancellation token source when there is more than one condition to cancel. For example, i suspect that here we need two conditions: 

1. The received cancellation token is cancelled **or**,
2. The time for waiting a round update has expired.

In that case we would have to do something like:

```c#
using var blameRoundWaitingCts = new CancellationTokenSource(TimeSpan.FromSeconds(timeToWaitForBlameRound));
using var effectiveCts = CancellationTokenSource.CreateLinkedTokenSource(blameRoundWaitingCts.Token,  cancellationToken);
var blameRoundState = await RoundStatusUpdater.CreateRoundAwaiter(..., effectiveCts.Token).....;

```

Reply:
ah I see what you mean now

Reply:
I think this is more or less correct, when the overall coinjoin is canceled, even if rounds have failed so far we no longer care about a blame round. Until then, to make the next attempt we need to find the right blame round.